### ISSUE TITLE
Analyzer plugin for .moor files

### ISSUE THREAD

**Issue by simolus3** [simolus3/moor#150]
> We can write an analyzer plugin for moor files that performs syntax highlighting, folding, outline views, errors, auto-complete, navigation and occurrences on moor files.

(I didn't see an existing issue for the analyzer plugin so feel free to move this as you see fit).
Navigation and syntax highlighting are the first features I'd personally see as most valuable. Being able to ctrl+click on generated dart methods and jump to the .moor file method and vice versa would be so cool - although I get the feeling that's probably the most challenging/complex feature!

_Originally posted by @Mike278 in https://github.com/simolus3/moor/issues/142#issuecomment-531575237_

**Comment by simolus3** [comment]
I've started some work on an analyzer plugin that lives in the `moor_generator` package. Sadly it can't really be previewed yet because no Dart IDE uses the results of the analysis server for custom file endings (the plugin does print some correct-looking outputs in the analysis instrumentation log though). I've opened https://github.com/Dart-Code/Dart-Code/issues/1981for that, so I hope this looks better in the next VS Code version. I'll open another issue for the IntelliJ plugin and reference that here as well.

### Syntax highlighting
Unfortunately, VS Code plugins can't provide dynamic syntax highlighting. They need to provide a static grammar to VS Code, so the Dart plugin can't use the moor analysis plugin for that. I looks like the IntelliJ plugin can the analysis server for highlighting, so maybe we have more luck there.

### Errors
We send both parsing and semantic lints, warnings and errors to the analysis server. Of course, this isn't nearly as helpful as syntax highlighting, but users would get a realtime feedback on invalid queries.

### Navigation and occurrences
To reduce complexity, I think it makes sense to first restrict this to sql in moor files only (so you could ctrl + click on a column name inside a query and jump to its definition). As long as the reference appears inside a moor file (and say the column is declared in a Dart file), there shouldn't be problems. Unfortunately, if a plugin sends a navigation event that is different from the Dart analyzer, the analyzer will take priority. So I'm not sure if we can find references to generated methods and "re-route" them to the moor file - the analyzer would probably veto those and link to the `.g.dart` file.

### Autocomplete
I've started some very basic work on this already. Given that we already have a parser, we can use that parsing logic to report an auto-complete option at a specific location. Of course, suggestions that depend on the context (e.g. column names, import files, etc.) would need special work.

### Folding and outline
This is arguably least spectacular feature, but at least it's already implemented. We provide information that lets editors collapse some definitions (like the column declarations in a `CREATE TABLE` statement or the import section). Outline gives an overview of the created tables, which columns they contain and so on in the side bar.

**Comment by simolus3** [comment]
With some [slight modifications](https://github.com/simolus3/moor/tree/develop/extras/plugin_example/README.md) of the Dart extension to VS Code, we can get outline, folding and errors:
![](https://user-images.githubusercontent.com/5738860/65168020-f8012780-da43-11e9-8ec9-8bb8671b3c23.png)

Syntax highlighting is blocked by VS Code, but we should be able to get auto-complete, quickfixes and navigation (inside moor files) working.

**Comment by simolus3** [comment]
Closing this as the prototype has landed in a stable release. Additions should be tracked in other issues.

**Comment by Mike278** [comment]
I discovered a decent workaround for IntelliJ users, but it requires Ultimate edition to have access to the "Database tools and SQL" plugin. 

1) Set the Recognized File Type for .moor files to SQL:
![Recognized File Type](https://i.imgur.com/EgfLN5p.png)

2) Write the database to a file. For example:
```dart
void main() {
  test('create db', () async {
    Database db = Database.withExecutor(VmDatabase(File('todos.sqlite3')));
    await db.categoriesWithCount().first;
  });
}
```
3) Add a the file as an sqlite datasource and connect:
![data source](https://i.imgur.com/zFOKWEC.png)

4) Configure the dialect for the directory as sqlite:
![dialect](https://i.imgur.com/R8079cF.png)

Result: 
![result](https://i.imgur.com/8ZidaWp.gif)
